### PR TITLE
[SPARK-18687][Pyspark][SQL]Backward compatibility - creating a Dataframe on a new SQLContext object fails with a Derby error

### PR DESCRIPTION
**Pull request by vijoshi**
## What changes were proposed in this pull request?

Change is for SQLContext to reuse the active SparkSession during construction if the sparkContext supplied is the same as the currently active SparkContext. Without this change, a new SparkSession is instantiated that results in a Derby error when attempting to create a dataframe using a new SQLContext object even though the SparkContext supplied to the new SQLContext is same as the currently active one. Refer https://issues.apache.org/jira/browse/SPARK-18687 for details on the error and a repro.

## How was this patch tested?

Existing unit tests and a new unit test added to pyspark-sql:

/python/run-tests --python-executables=python --modules=pyspark-sql

Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

**Comment by vijoshi** [comment]
ping @holdenk @srowen

**Comment by holdenk** [comment]
Since the current tests pass without this change I'd say that we should add a test for the behaviour we are planning to support that isn't currently supported (would also make the purpose of the change a bit clearer).

**Comment by vijoshi** [comment]
@holdenk test case added

**Comment by vijoshi** [comment]
@holdenk reminder 
also pinging @davies

File: python/pyspark/sql/context.py [hunk 1]
```diff
@@ -72,8 +72,13 @@ def __init__(self, sparkContext, sparkSession=None, jsqlContext=None):
         self._sc = sparkContext
         self._jsc = self._sc._jsc
         self._jvm = self._sc._jvm
+
         if sparkSession is None:
-            sparkSession = SparkSession(sparkContext)
+            if sparkContext is SparkContext._active_spark_context:
+                sparkSession = SparkSession.builder.getOrCreate()
```
Comment (srowen) [inline python/pyspark/sql/context.py]: I don't know enough to usefully review, but is the better fix to call the getOrCreate that lets you pass a context and use that for both cases? I don't know if it is the same behavior but seems more natural to use the logic in SparkSession for this.

Comment (holdenk) [inline python/pyspark/sql/context.py]: Thats a good question, lets see what @andrewor14 thinks of changing the builder interface for `SparkSession`

Comment (srowen) [inline python/pyspark/sql/context.py]: @viirya since you opened a similar change, what do you think of this?

Comment (viirya) [inline python/pyspark/sql/context.py]: So you can see in the pr, I simply do `sparkSession = SparkSession.builder.getOrCreate()` here.

File: python/pyspark/sql/tests.py [hunk 2]
```diff
@@ -1851,6 +1851,71 @@ def test_hivecontext(self):
         self.assertIn("default", out.decode('utf-8'))
         self.assertTrue(os.path.exists(metastore_path))
 
+    def test_hivecontext_create_dataframe(self):
```
Comment (cloud-fan) [inline python/pyspark/sql/tests.py]: can we write a more low-level unit test to check if we create 2 `SparkSession`s?

File: python/pyspark/sql/context.py [hunk 3]
```diff
@@ -72,8 +72,13 @@ def __init__(self, sparkContext, sparkSession=None, jsqlContext=None):
...
+            else:
```
Comment (cloud-fan) [inline python/pyspark/sql/context.py]: when will hit the else branch? The if-else seems the only difference between this PR and https://github.com/apache/spark/pull/16369

File: python/pyspark/sql/context.py (hunk 1 above)
Comment (vijoshi) [inline python/pyspark/sql/context.py]: 1. the `SQLContext __init__` suggested that any sparkSession object could be injected. And the `SparkSession __init__` suggested any sparkContext could be injected. If not so - I thought they would all be invoking `getOrCreate()` to get instances of each other.  

2.  I found a few hits around this property - `spark.driver.allowMultipleContexts` which though not a python property - showed that there may be cases where someone (for testing etc) wanted the ability to bypass the one sparkContext limitation. 

These two combined I was not 100% sure someone couldn't send in different sparkContext objects / sparkSession objects tied to different sparkContexts when constructing an SQLContext. 

I will admit I am not familiar with Python side of things, so I hoped this PR review would be the best way to get this clarified :). 

If everyone agrees that we needn't worry about what sparkSession is passed in when constructing SQLContext, then I can get rid of the `if sparkContext is SparkContext._active_spark_context:` check.

File: python/pyspark/sql/context.py (hunk 3 above)
Comment (vijoshi) [inline python/pyspark/sql/context.py]: @cloud-fan refer my comments above.

File: python/pyspark/sql/context.py (hunk 1 above)
Comment (viirya) [inline python/pyspark/sql/context.py]: I found a relate discussion about multiple `SparkContext` at http://stackoverflow.com/questions/28259756/how-to-create-multiple-sparkcontexts-in-a-console

I think the usage of multiple contexts is discouraged, even you could do it in Scala/Java. Specially, you can't have multiple `SparkContext` in Python. So I am not worrying that another `SparkContext` other than the active one is passed in when constructing `SQLContext`.

File: python/pyspark/sql/tests.py (hunk 2 above)
Comment (vijoshi) [inline python/pyspark/sql/tests.py]: did you mean have the same scenario where multiple SQLContext objects are constructed but the test success criteria is to verify that no more than a single `SparkSession` object ever gets created?

Comment (viirya) [inline python/pyspark/sql/tests.py]: I think we can just create two `SQLContext` and compare if they share the same `SparkSession` instance.

**Comment by cloud-fan** [comment]
@vijoshi do you mind updating your PR according to the dicussion? i.e. simplify the fix and test

**Comment by vijoshi** [comment]
@cloud-fan sure, i'll push an update soon

**Comment by vijoshi** [comment]
@cloud-fan @viirya - rebased and updated the code change per review comments.

**Comment by vijoshi** [comment]
retest this please

**Comment by viirya** [comment]
The test failure is caused by another issue. I've submitted another PR to fix it: #16454. Once that is fixed, this test can be passed.

**Comment by vijoshi** [comment]
thanks @viirya. will wait for your PR to go through.

**Comment by cloud-fan** [comment]
https://github.com/apache/spark/pull/16454 has been merged

**Comment by viirya** [comment]
@vijoshi Can you solve the conflict?

**Comment by cloud-fan** [comment]
thanks, merging to master/2.1/2.0!

**Pull request closed without merging**